wp_api/src/login/nonce.rs

```diff
@@ -0,0 +1,183 @@
+use std::sync::Arc;
+
+use http::HeaderValue;
+use url::Url;
+use uuid::Uuid;
+use wp_localization::{WpMessages, WpSupportsLocalization};
+use wp_localization_macro::WpDeriveLocalizable;
+
+use crate::{
+    EmptyAppNotifier,
+    login::url_discovery::AutoDiscoveryAttemptSuccess,
+    prelude::*,
+    request::{
+        RequestMethod, WpNetworkRequestBody, endpoint::users_endpoint::UsersRequestExecutor,
+    },
+};
+
+#[derive(uniffi::Object)]
+pub struct WpRestNonceRetrieval {
+    details: AutoDiscoveryAttemptSuccess,
+    request_executor: Arc<dyn RequestExecutor>,
+}
+
+#[uniffi::export]
+impl WpRestNonceRetrieval {
+    #[uniffi::constructor]
+    pub fn new(
+        details: AutoDiscoveryAttemptSuccess,
+        request_executor: Arc<dyn RequestExecutor>,
+    ) -> Self {
+        Self {
+            details,
+            request_executor,
+        }
+    }
+
+    pub async fn get_nonce(
+        &self,
+        username: String,
+        password: String,
+    ) -> Result<String, NonceRetrievalError> {
+        let mut nonce = self.nonce_from_request(self.nonce_request()).await;
+
+        if let Err(NonceRetrievalError::UnexpectedResponse { .. }) = nonce {
+            nonce = self
+                .nonce_from_request(self.nonce_request_via_login(&username, &password))
+                .await;
+        }
+
+        // Since the "cookies" part is out of our control, we need to verify that the nonce we got
+        // is actually for the user we expect.
+        if let Ok(nonce) = nonce.as_ref() {
+            let api_url = WpOrgSiteApiUrlResolver::new(self.details.api_root_url.clone());
+            let auth = WpAuthentication::Nonce {
+                nonce: nonce.clone(),
+            };
+            let users = UsersRequestExecutor::new(
+                Arc::new(api_url),
+                WpApiClientDelegate {
+                    auth_provider: WpAuthenticationProvider::static_with_auth(auth).into(),
+                    request_executor: self.request_executor.clone(),
+                    middleware_pipeline: WpApiMiddlewarePipeline::default().into(),
+                    app_notifier: Arc::new(EmptyAppNotifier),
+                },
+            );
+            let logged_in = users.retrieve_me_with_edit_context().await?.data.username;
+            if logged_in != username {
+                return Err(NonceRetrievalError::AlreadyLoggedIn {
+                    username: logged_in,
+                });
+            }
+        }
+
+        nonce
+    }
+}
+
+impl WpRestNonceRetrieval {
+    fn derived_login_url(&self) -> Url {
+        let mut url = self.details.parsed_site_url.inner.clone();
+        url.path_segments_mut()
+            .expect("The site url is a full URL")
+            .push("wp-login.php");
+        url
```

Comment:
You should be able to use:


```suggestion
        self.details
            .application_passwords_authentication_url
            .inner
            .clone()
```